### PR TITLE
Websokets Connection

### PR DESCRIPTION
**Pull request by silicon-ninja**
You can try all the methods from localhost/docs for a simple schema 
1. Get 
2. Post 
3. Delete 
----------------
For my next PR I will be redoing the final schema and updating it with Web-socket Connection

### CODE REVIEW

**Comment by firekind** [comment]
Ok problems (which you might already know):
1. The template is not followed. all your code is there in the `database` package.
2. The APIs you programmed are not even needed.
3. Good attempt at documentation, but use [this](https://marketplace.visualstudio.com/items?itemName=njpwerner.autodocstring) extension to create proper doc strings. also, remove all the `\n`s

File: app/api/dbroutes.py [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
 from pydantic import BaseModel
-some_router = APIRouter() 
 
-@some_router.get('/findallmessage')
-async def list_message():
+some_router = APIRouter()
+
+
+@some_router.get('/message/get')
+async def list_message(channel_id:str,starttime:str,endtime:str):
```
Comment (shikharx06) [inline app/api/dbroutes.py]: use snake_case only

File: app/api/dbroutes.py [hunk 2]
```diff
@@ -16,12 +19,14 @@ async def list_message():
     Return Type : List 
     
     """
-    return get_message()
-  
+    return get_message(channel_id,starttime,endtime)
```
Comment (shikharx06) [inline app/api/dbroutes.py]: snake_case!!

File: app/api/dbroutes.py [hunk 3]
```diff
@@ -16,12 +19,14 @@ async def list_message():
...
+
+
 
-@some_router.post('/createmessage')
-async def post_message(message:Message) :
-   """
+
+@some_router.post('/message/post')
+async def post_message(message: Messagebody):
```
Comment (shikharx06) [inline app/api/dbroutes.py]: snake_case!!! And write proper docstrings with suitable description followed by `args`, `returns` and `raises`. No need to use `\n` anywhere

File: app/database/controller.py [hunk 4]
```diff
@@ -1,54 +1,66 @@
...
+from bson import ObjectId 
+import datetime 
+from tzlocal import get_localzone # $ pip install tzlocal
 
-def get_message():
+
+some_router = APIRouter()
+
+
+def get_message(channel_id,starttime,endtime):
```
Comment (shikharx06) [inline app/database/controller.py]: docstrings!!!!

File: app/database/controller.py [hunk 5]
```diff
@@ -1,54 +1,66 @@
...
     
     The Function is used to fetch all the data in the DB 
     
     Return Type : List 
-    
+    {"channel_id":channel_id })
     """
+    starttime = dateutil.parser.parse(starttime)
+    endtime = dateutil.parser.parse(endtime)
+    print(starttime)
```
Comment (shikharx06) [inline app/database/controller.py]: remove print statements before raising a PR.

File: app/database/controller.py [hunk 6]
```diff
@@ -1,54 +1,66 @@
...
+    print(endtime)
     cynergy = []
-    print (db.message.find())
```
Comment (shikharx06) [inline app/database/controller.py]: create separate file containing DB methods and call appropriate method whenever needed. Instead of writing `db.message.find()` you should write `message_db().find()`. Do this everywhere!

File: app/database/controller.py [hunk 7]
```diff
@@ -1,54 +1,66 @@
...
 
 
-def create_message(name:str,message:str):
-    """
-    Args: 
-    \n Name -> str
-    \n Message -> str
-    
+def create_message(channel_id:str,content:str,sender_id:str):
+    """[summary]
```
Comment (shikharx06) [inline app/database/controller.py]: unpolished code.  😟

File: app/database/controller.py [hunk 8]
```diff
@@ -1,54 +1,66 @@
...
-    return {"Name":name,"Message":message}
+    Args:
+        channel_id (str): [description]
+        content (str): [description]
+        sender_id (str): [description]
+        timestamp (str): [description]
 
-def remove_message(_id):
+    Returns:
+        [List]: [description]
```
Comment (shikharx06) [inline app/database/controller.py]: list of what? if it is list of mesages, it should be `List[messages]`. also write appropriate description.

File: app/database/controller.py [hunk 9]
```diff
@@ -1,54 +1,66 @@
...
     """
-    Args: 
-    \n _id -> String 
+    timestamp = datetime.datetime.now()
+    ret = db.message.insert_one({"channel_id":channel_id,"content":content,"sender_id":sender_id,"timestamp":timestamp})
```
Comment (shikharx06) [inline app/database/controller.py]: always wrap db related update and insertion commands with `try/except` and catch appropriate pymongo error whenever and whereever needed. you should also raise custom errors wherever you seem its necessary to do so and mention the same in docstrings.

File: app/database/schema.py [hunk 10]
```diff
@@ -28,29 +29,18 @@ def validate(cls, v):
...
-    
-     Updated Schema for next week:
-    
-        0. Object_id
-        1. sender_id
-        2. channel_id
-        3.message_body
-        4.time_stamp
-    
+    """[summary]
```
Comment (shikharx06) [inline app/database/schema.py]: there's no point in giving docstrings if you are going to leave it like this. Better habit would be to write doc strings first then proceed with writing the code. It gives you much better clarity of your logic of what you will do.

File: app/services/wsconfig.py [hunk 11]
```diff
@@ -1,49 +1,55 @@
...
+            }
+        </script>
+    </body>
+</html>
+"""
+
+
+@webapp.get("/")
+async def get():
+    return HTMLResponse(html)
```
Comment (shikharx06) [inline app/services/wsconfig.py]: why do you have to return the html? frontend people only need the following things:
 - channel_id
 - sender_id
 - message
 - timestamp
 - url of the image if it's an image
 
Do a complete review of your code before raising a PR or do a peer review first, gather inputs from others then raise a PR.

File: app/services/wsconfig.py [hunk 12]
```diff
@@ -1,49 +1,55 @@
...
+
+
+@webapp.websocket("/ws")
+async def websocket_endpoint(websocket: WebSocket):
+    await websocket.accept()
+    while True:
+        data = await websocket.receive_text()
+        await websocket.send_text(f"Message text was: {data}")
```
Comment (shikharx06) [inline app/services/wsconfig.py]: only need to send data.

**Review by shikharx06 — DISMISSED** [review]

**Comment by silicon-ninja** [comment]
Hey @firekind and @shikharx06 WebSockets work properly.
 Please check

File: Dockerfile [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+COPY requirements.txt /opt/app/requirements.txt
+WORKDIR /opt/app
+RUN pip install -r requirements.txt
+
+RUN groupadd prod && useradd -g prod prod
+USER prod:prod
+
+COPY ${SRC} app
+COPY ${SRC}/../templates templates
```
Comment (firekind) [inline Dockerfile]: remove this line from the docker file, that way the test frontend wont be available in production

File: app/api/routes.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    This route is used to get the messages from the database and is handled by a controller
+    get_message() from message_service.py
+
+    Args:
+        channel_id (str): The channel UUID
+        start_time (str): The start time
+        end_time (str): The end time
+
+    Returns:
+        [List]: A List of all the messages from the channel between two times
```
Comment (firekind) [inline app/api/routes.py]: remove square brackets

File: app/api/websockets_routes.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+@websoc.on_event("startup")
+async def startup_event():
+    """
+    This function trigers to send the messages through sockets :
+    """
+    await notifier.generator.asend(None)
+
+
+@websoc.get("/test/{channel_id}/{sender_id}")
+async def get(request: Request, channel_id, sender_id):
```
Comment (firekind) [inline app/api/websockets_routes.py]: type hints missing

File: app/api/websockets_routes.py [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+        returns the response in HTML file with websockets in frontend :
+    """
+    return templates.TemplateResponse(
+        "index.html",
+        {"request": request, "channel_id": channel_id, "sender_id": sender_id},
+    )
+
+
+@websoc.websocket("/ws/{channel_id}/{sender_id}")
+async def websocket_endpoint(websocket: WebSocket, channel_id):
```
Comment (firekind) [inline app/api/websockets_routes.py]: type hints missing

File: app/database/database_queries.py [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+                "$and": [
+                    {"timestamp": {"$gte": start_time}, "timestamp": {"$lt": end_time}}
+                ],
+            }
+        )
+    )
+
+
+async def create_message(channel_id: str, content: str, sender_id: str):
+    """[summary]
```
Comment (firekind) [inline app/database/database_queries.py]: remove the `[summary]` thing

File: app/services/message_service.py [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+        List: A List of all the messages from the channel between two times
+    """
+    try:
+        start_time = dateutil.parser.parse(start_time)
+        end_time = dateutil.parser.parse(end_time)
+        if not mongo_queries.check_channel_exists(channel_id):
+            raise APIErrors(404, "The channel you are looking for doesn't exist")
+        return mongo_queries.get_message_between(channel_id, start_time, end_time)
+    except pymongo.errors.CollectionInvalid as e:
+        logging.info(e)
```
Comment (firekind) [inline app/services/message_service.py]: 1. change `logging.info(e)` to `logging.error(e)`
2. after the logging statement, raise an `APIErrors` with error code 500.

File: app/services/message_service.py [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+        timestamp = datetime.datetime.now()
+        mongo_queries.create_message(channel_id, content, sender_id, timestamp)
+        return {
+            "channel_id": channel_id,
+            "content": content,
+            "sender_id": sender_id,
+            "timestamp": timestamp,
+        }
+    except pymongo.errors.CollectionInvalid as e:
+        logging.info(e)
```
Comment (firekind) [inline app/services/message_service.py]: 1. change `logging.info(e)` to `logging.error(e)`
2. after the logging statement, raise an `APIErrors` with error code 500

**Comment by firekind** [comment]
Found a bug. Look at this pic:



the message sent in one channel goes to the other channels also.

**Comment by silicon-ninja** [comment]
> Found a bug. Look at this pic:
> 
> 
> 
> the message sent in one channel goes to the other channels also.

this is because I didn't change the url extension in frontend Websocket connection so that happens I will change that

File: app/services/message_service.py [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+        end_time = dateutil.parser.parse(end_time)
+        if not mongo_queries.check_channel_exists(channel_id):
+            raise APIErrors(404, "The channel you are looking for doesn't exist")
+        return mongo_queries.get_message_between(channel_id, start_time, end_time)
+    except pymongo.errors.CollectionInvalid as e:
+        logging.error(e)
+    except ParserError as e:
+        raise APIErrors(422, "Invalid Time Format.")
+    except logging.error(e) as e:
+        raise APIErrors(500," Mongo Internal Server Error ")
```
Comment (firekind) [inline app/services/message_service.py]: not like this :facepalm: remove this part

File: app/services/message_service.py [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+        List: A List of all the messages from the channel between two times
+    """
+    try:
+        start_time = dateutil.parser.parse(start_time)
+        end_time = dateutil.parser.parse(end_time)
+        if not mongo_queries.check_channel_exists(channel_id):
+            raise APIErrors(404, "The channel you are looking for doesn't exist")
+        return mongo_queries.get_message_between(channel_id, start_time, end_time)
+    except pymongo.errors.CollectionInvalid as e:
+        logging.error(e)
```
Comment (firekind) [inline app/services/message_service.py:38]: change this line to:
```python
logging.error(e)
raise APIErrors(500, "Mongo internal server error")
```
```

File: app/services/message_service.py [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+        timestamp = datetime.datetime.now()
+        mongo_queries.create_message(channel_id, content, sender_id, timestamp)
+        return {
+            "channel_id": channel_id,
+            "content": content,
+            "sender_id": sender_id,
+            "timestamp": timestamp,
+        }
+    except pymongo.errors.CollectionInvalid as e:
+        logging.info(e)
```
Comment (firekind) [inline app/services/message_service.py]: same here also

File: app/services/message_service.py [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+    except logging.error(e) as e:
+        raise APIErrors(500," Mongo Internal Server Error ")
```
Comment (firekind) [inline app/services/message_service.py]: and remove this

File: app/config.py [hunk 24]
```diff
@@ -0,0 +1,5 @@
+import os
+from decouple import config
+MONGO_URL = config('MONGO_URL')
+DB_NAME = config('DB_NAME')
+COLLECTION_NAME = config('COLLECTION_NAME')
```
Comment (firekind) [inline app/config.py]: wait wth is this? it was fine before

Comment (silicon-ninja) [reply app/config.py]: oh yeah I looked it up online this is like the alternative way of using environmental variables

File: app/api/websockets_routes.py [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+
+@websoc.on_event("startup")
+async def startup_event():
+    """
+    This function trigers to send the messages through sockets :
+    """
+    await notifier.generator.asend(None)
+
+
+@websoc.get("/{channel_id}/{sender_id}")
```
Comment (firekind) [inline app/api/websockets_routes.py:24]: and why did you remove `test` from the path?

File: app/config.py (hunk 24 above)
Comment (firekind) [reply app/config.py]: doesnt work (you probably forgot to put decouple in `requirements.txt`). Switch to regular `os.environ`

Comment (silicon-ninja) [reply app/config.py]: okay then sure

File: README.md [hunk 26]
```diff
@@ -1,2 +1,106 @@
...
+```
+And you can run the server using
+```
+$ python start.py
+``` 
+
+Windows:
+```
+$ MONGO_URL= <Mongo_URL> DB_NAME=<Database_Name> COLLECTION_NAME=<Message_Collection_Name> python start.py
+```
```
Comment (firekind) [inline README.md]: this doesnt work like this in windows..in powershell, its
```
$env:MONGO_URL=<mongo_url>
$env:DB_NAME=<db_name>
$env:COLLECTION_NAME=<collection_name>
./venv/Scripts/activate # activate the environment
python start.py
```

File: README.md [hunk 27]
```diff
@@ -1,2 +1,106 @@
...
+     -e DB_NAME=<Database_Name> \
+     -e cynergyruas/ruas-app:message-service-0.0.1  
+```
+or you can run the server using 
+
+Linux:
+
+To export the local environment varibles 
+```
+$ export $(cat .env | xargs)
```
Comment (firekind) [inline README.md:99]: explain what should be there in the `.env` file. and this line is linux only, not windows

**Comment by firekind** [comment]
Alright! good work!

**Review by firekind — APPROVED** [review]
